src/ppx/xr_component.cpp

```diff
@@ -0,0 +1,395 @@
+    VERIFY_XR_CALL(xrCreateInstance(&createInfo, &mInstance));
+    PPX_ASSERT_MSG(mInstance != nullptr, "XrInstance Creation Failed!");
+
+    // Get System Id
+    XrSystemGetInfo systemInfo = {XR_TYPE_SYSTEM_GET_INFO};
+    systemInfo.formFactor      = XR_FORM_FACTOR_HEAD_MOUNTED_DISPLAY;
+    VERIFY_XR_CALL(xrGetSystem(mInstance, &systemInfo, &mSystemId));
+
+    // Use the first available Blend Mode
+    uint32_t blendCount = 0;
+    VERIFY_XR_CALL(xrEnumerateEnvironmentBlendModes(mInstance, mSystemId, mViewConfigType, 1, &blendCount, &mBlend));
+
+    // Create Debug Utils Messenger
+    if (enableDebug) {
+        XrDebugUtilsMessengerCreateInfoEXT debug_info = {XR_TYPE_DEBUG_UTILS_MESSENGER_CREATE_INFO_EXT};
+        debug_info.messageTypes =
+            XR_DEBUG_UTILS_MESSAGE_TYPE_GENERAL_BIT_EXT |
+            XR_DEBUG_UTILS_MESSAGE_TYPE_VALIDATION_BIT_EXT |
+            XR_DEBUG_UTILS_MESSAGE_TYPE_PERFORMANCE_BIT_EXT |
+            XR_DEBUG_UTILS_MESSAGE_TYPE_CONFORMANCE_BIT_EXT;
+        debug_info.messageSeverities =
+            XR_DEBUG_UTILS_MESSAGE_SEVERITY_VERBOSE_BIT_EXT |
+            XR_DEBUG_UTILS_MESSAGE_SEVERITY_INFO_BIT_EXT |
+            XR_DEBUG_UTILS_MESSAGE_SEVERITY_WARNING_BIT_EXT |
+            XR_DEBUG_UTILS_MESSAGE_SEVERITY_ERROR_BIT_EXT;
+        debug_info.userCallback = [](XrDebugUtilsMessageSeverityFlagsEXT severity, XrDebugUtilsMessageTypeFlagsEXT types, const XrDebugUtilsMessengerCallbackDataEXT* msg, void* user_data) {
+            printf("%s: %s\n", msg->functionName, msg->message);
+            // Returning XR_TRUE here will force the calling function to fail
+            return (XrBool32)XR_FALSE;
+        };
+
+        PFN_xrCreateDebugUtilsMessengerEXT pfnCreateDebugUtilsMessengerEXT = nullptr;
+        VERIFY_XR_CALL(xrGetInstanceProcAddr(mInstance, "xrCreateDebugUtilsMessengerEXT", (PFN_xrVoidFunction*)(&pfnCreateDebugUtilsMessengerEXT)));
+        PPX_ASSERT_MSG(pfnCreateDebugUtilsMessengerEXT != nullptr, "Cannot get xrCreateDebugUtilsMessengerEXT function pointer!");
+        VERIFY_XR_CALL(pfnCreateDebugUtilsMessengerEXT(mInstance, &debug_info, &mDebugUtilMessenger));
+    }
+}
+
+void XrComponent::InitializeAfterGrfxDeviceInit(const grfx::InstancePtr pGrfxInstance)
+{
+    PPX_ASSERT_MSG(!pGrfxInstance.IsNull(), "Invalid Instance!");
+    const grfx::Instance& grfxInstance = *pGrfxInstance;
+    PPX_ASSERT_MSG(grfxInstance.IsGraphicsBindingValid(), "Invalid Graphics Binding!");
+
+    // Create XrSession
+    XrSessionCreateInfo sessionInfo = {XR_TYPE_SESSION_CREATE_INFO};
+    sessionInfo.next                = grfxInstance.GetGraphicsBinding();
+    sessionInfo.systemId            = mSystemId;
+    VERIFY_XR_CALL(xrCreateSession(mInstance, &sessionInfo, &mSession));
+    PPX_ASSERT_MSG(mSession != XR_NULL_HANDLE, "XrSession creation failed!");
+
+    // Create Reference Space
+    // TODO(wangra): need to decide if we should use Local or Stage
```

Comment:
Another good thing to pass in the constructor

Reply:
fixed